File: EvolutionaryGenerator_v2.1_Java/Generator/src/mx/tec/hermes/problems/kp/generator/KPEvaluator.java
package mx.tec.hermes.problems.kp.generator;

import mx.tec.meta.Evaluator;
import mx.tec.meta.Individual;
import mx.tec.hermes.utils.Statistical;
import mx.tec.hermes.problems.kp.KP;

/**
 * Provides the methods to evaluate individuals for the automatic generation of knapsack problems by using a genetic
 * algorithm.
 *
 * @author José Carlos Ortiz Bayliss (jcobayliss@tec.mx)
 * @version 2.0
 */
public class KPEvaluator extends Evaluator {
    
    private final InstanceType type;

    public enum InstanceType {
        DEFAULT_EASY,
        DEFAULT_HARD,
        MIN_WEIGHT_EASY,
        MIN_WEIGHT_HARD,
        MAX_PROFIT_EASY,
        MAX_PROFIT_HARD,
        MAX_PROFIT_WEIGHT_EASY,
        MAX_PROFIT_WEIGHT_HARD,
        MARKOVITZ_EASY,
        MARKOVITZ_HARD,
        MIN_VARIANCE,
        MAX_VARIANCE,
        PAIRED_DEF_MAXP,
        PAIRED_DEF_MAXPW,
        PAIRED_DEF_MINW,
        PAIRED_MAXP_MAXPW,
        PAIRED_MAXP_MINW,
        PAIRED_MAXPW_MINW
    }

    /**
     * Creates a new instance of <code>KPEvaluator</code>.
     *
     * @param type The type of the instances to produce.
     */
    public KPEvaluator(InstanceType type) {
        super();
        this.type = type;
    }
   
    @Override
    public double evaluate(Individual individual) {
        double resultDefault, resultMinWeight, resultMaxProfitPerWeightUnit, resultMaxProfit, max, lambda;
        double[] results;
        KP problem;
        nbEvaluations++;
        problem = ((KPIndividual) individual).toKP();
        problem.solve("DEFAULT");
        resultDefault = problem.getObjValue();
        problem = ((KPIndividual) individual).toKP();
        problem.solve("MIN_WEIGHT");
        resultMinWeight = problem.getObjValue();
        problem = ((KPIndividual) individual).toKP();
        problem.solve("MAX_PROFIT/WEIGHT");
        resultMaxProfitPerWeightUnit = problem.getObjValue();
        problem = ((KPIndividual) individual).toKP();
        problem.solve("MAX_PROFIT");
        resultMaxProfit = problem.getObjValue();
        lambda = 2;
        switch (type) {
            case DEFAULT_EASY:
                results = new double[]{resultMinWeight, resultMaxProfit, resultMaxProfitPerWeightUnit};
                return resultDefault - Statistical.max(results);
            case DEFAULT_HARD:
                results = new double[]{resultMinWeight, resultMaxProfit, resultMaxProfitPerWeightUnit};
                return Statistical.min(results) - resultDefault;
            case MIN_WEIGHT_EASY:
                results = new double[]{resultDefault, resultMaxProfit, resultMaxProfitPerWeightUnit};
                return resultMinWeight - Statistical.max(results);
            case MIN_WEIGHT_HARD:
                results = new double[]{resultDefault, resultMaxProfit, resultMaxProfitPerWeightUnit};
                return Statistical.min(results) - resultMinWeight;
            case MAX_PROFIT_EASY:
                results = new double[]{resultDefault, resultMinWeight, resultMaxProfitPerWeightUnit};
                return resultMaxProfit - Statistical.max(results) ;
            case MAX_PROFIT_HARD:
                results = new double[]{resultDefault, resultMinWeight, resultMaxProfitPerWeightUnit};
                return Statistical.min(results) - resultMaxProfit;
            case MAX_PROFIT_WEIGHT_EASY:
                results = new double[]{resultDefault, resultMinWeight, resultMaxProfit};
                return resultMaxProfitPerWeightUnit - Statistical.max(results);
            case MAX_PROFIT_WEIGHT_HARD:
                results = new double[]{resultDefault, resultMinWeight, resultMaxProfit};
                return Statistical.min(results) - resultMaxProfitPerWeightUnit;
            case MAX_VARIANCE:
                results = new double[]{resultDefault, resultMinWeight, resultMaxProfit, resultMaxProfitPerWeightUnit};
                return Statistical.stdev(results);
            case MIN_VARIANCE:
                results = new double[]{resultDefault, resultMinWeight, resultMaxProfit, resultMaxProfitPerWeightUnit};
                return -1 * Statistical.stdev(results);
            case PAIRED_DEF_MAXP:
                max = Statistical.max(new double[]{resultDefault, resultMaxProfit, resultMaxProfitPerWeightUnit, resultMinWeight});
                resultDefault = resultDefault / max;
                resultMaxProfit = resultMaxProfit / max;
                resultMaxProfitPerWeightUnit = resultMaxProfitPerWeightUnit / max;
                resultMinWeight = resultMinWeight / max;
                return Math.min(resultDefault, resultMaxProfit) - Math.max(resultMaxProfitPerWeightUnit, resultMinWeight) - lambda * Math.abs(resultDefault - resultMaxProfit);            
            case PAIRED_DEF_MAXPW:
                max = Statistical.max(new double[]{resultDefault, resultMaxProfit, resultMaxProfitPerWeightUnit, resultMinWeight});
                resultDefault = resultDefault / max;
                resultMaxProfit = resultMaxProfit / max;
                resultMaxProfitPerWeightUnit = resultMaxProfitPerWeightUnit / max;
                resultMinWeight = resultMinWeight / max;
                return Math.min(resultDefault, resultMaxProfitPerWeightUnit) - Math.max(resultMaxProfit, resultMinWeight) - lambda * Math.abs(resultDefault - resultMaxProfitPerWeightUnit);            
            case PAIRED_DEF_MINW:
                max = Statistical.max(new double[]{resultDefault, resultMaxProfit, resultMaxProfitPerWeightUnit, resultMinWeight});
                resultDefault = resultDefault / max;
                resultMaxProfit = resultMaxProfit / max;
                resultMaxProfitPerWeightUnit = resultMaxProfitPerWeightUnit / max;
                resultMinWeight = resultMinWeight / max;
                return Math.min(resultDefault, resultMinWeight) - Math.max(resultMaxProfit, resultMaxProfitPerWeightUnit) - lambda * Math.abs(resultDefault - resultMinWeight);            
            case PAIRED_MAXP_MAXPW:
                max = Statistical.max(new double[]{resultDefault, resultMaxProfit, resultMaxProfitPerWeightUnit, resultMinWeight});
                resultDefault = resultDefault / max;
                resultMaxProfit = resultMaxProfit / max;
                resultMaxProfitPerWeightUnit = resultMaxProfitPerWeightUnit / max;
                resultMinWeight = resultMinWeight / max;
                return Math.min(resultMaxProfit, resultMaxProfitPerWeightUnit) - Math.max(resultDefault, resultMinWeight) - lambda * Math.abs(resultMaxProfit - resultMaxProfitPerWeightUnit);            
            case PAIRED_MAXP_MINW:
                max = Statistical.max(new double[]{resultDefault, resultMaxProfit, resultMaxProfitPerWeightUnit, resultMinWeight});
                resultDefault = resultDefault / max;
                resultMaxProfit = resultMaxProfit / max;
                resultMaxProfitPerWeightUnit = resultMaxProfitPerWeightUnit / max;
                resultMinWeight = resultMinWeight / max;
                return Math.min(resultMaxProfit, resultMinWeight) - Math.max(resultDefault, resultMaxProfitPerWeightUnit) - lambda * Math.abs(resultMaxProfit - resultMinWeight);
            case PAIRED_MAXPW_MINW:                
                max = Statistical.max(new double[]{resultDefault, resultMaxProfit, resultMaxProfitPerWeightUnit, resultMinWeight});
                resultDefault = resultDefault / max;
                resultMaxProfit = resultMaxProfit / max;
                resultMaxProfitPerWeightUnit = resultMaxProfitPerWeightUnit / max;
                resultMinWeight = resultMinWeight / max;
                return Math.min(resultMaxProfitPerWeightUnit, resultMinWeight) - Math.max(resultDefault, resultMaxProfit) - lambda * Math.abs(resultMaxProfitPerWeightUnit - resultMinWeight);
            default:
                System.err.println("The option is not recognized by the system.");
                System.err.println("The system will halt.");
                System.exit(1);
        }
        return 0;       
    }

}

File: EvolutionaryGenerator_v2.1_Python/hermes/problems/kp/generator/kp_evaluator.py
"""Translation of ``mx.tec.hermes.problems.kp.generator.KPEvaluator``."""

import sys
from enum import Enum

from hermes.utils import statistical as Statistical
from metah.evaluator import Evaluator
from metah.individual import Individual


class InstanceType(Enum):
    """Defines the type of the instances the evaluator asks the genetic algorithm to produce."""

    DEFAULT_EASY = "DEFAULT_EASY"
    DEFAULT_HARD = "DEFAULT_HARD"
    MIN_WEIGHT_EASY = "MIN_WEIGHT_EASY"
    MIN_WEIGHT_HARD = "MIN_WEIGHT_HARD"
    MAX_PROFIT_EASY = "MAX_PROFIT_EASY"
    MAX_PROFIT_HARD = "MAX_PROFIT_HARD"
    MAX_PROFIT_WEIGHT_EASY = "MAX_PROFIT_WEIGHT_EASY"
    MAX_PROFIT_WEIGHT_HARD = "MAX_PROFIT_WEIGHT_HARD"
    MARKOVITZ_EASY = "MARKOVITZ_EASY"
    MARKOVITZ_HARD = "MARKOVITZ_HARD"
    MIN_VARIANCE = "MIN_VARIANCE"
    MAX_VARIANCE = "MAX_VARIANCE"
    PAIRED_DEF_MAXP = "PAIRED_DEF_MAXP"
    PAIRED_DEF_MAXPW = "PAIRED_DEF_MAXPW"
    PAIRED_DEF_MINW = "PAIRED_DEF_MINW"
    PAIRED_MAXP_MAXPW = "PAIRED_MAXP_MAXPW"
    PAIRED_MAXP_MINW = "PAIRED_MAXP_MINW"
    PAIRED_MAXPW_MINW = "PAIRED_MAXPW_MINW"


class KPEvaluator(Evaluator):
    """Provides the methods to evaluate individuals for the automatic generation of
    knapsack problems by using a genetic algorithm.

    :author: Paola Azeneth Castillo Gutiérrez (Python translation)
    :original author: José Carlos Ortiz Bayliss (jcobayliss@tec.mx)
    :original version: 2.0
    """

    InstanceType = InstanceType

    def __init__(self, type: InstanceType):
        """Creates a new instance of ``KPEvaluator``.

        :param type: The type of the instances to produce.
        """
        super().__init__()
        self.type = type

    def evaluate(self, individual: Individual) -> float:
        self.nb_evaluations += 1
        problem = individual.to_kp()
        problem.solve("DEFAULT")
        result_default = problem.get_obj_value()
        problem = individual.to_kp()
        problem.solve("MIN_WEIGHT")
        result_min_weight = problem.get_obj_value()
        problem = individual.to_kp()
        problem.solve("MAX_PROFIT/WEIGHT")
        result_max_profit_per_weight_unit = problem.get_obj_value()
        problem = individual.to_kp()
        problem.solve("MAX_PROFIT")
        result_max_profit = problem.get_obj_value()
        lambda_ = 2
        type = self.type
        if type == InstanceType.DEFAULT_EASY:
            results = [result_min_weight, result_max_profit, result_max_profit_per_weight_unit]
            return result_default - Statistical.max(results)
        elif type == InstanceType.DEFAULT_HARD:
            results = [result_min_weight, result_max_profit, result_max_profit_per_weight_unit]
            return Statistical.min(results) - result_default
        elif type == InstanceType.MIN_WEIGHT_EASY:
            results = [result_default, result_max_profit, result_max_profit_per_weight_unit]
            return result_min_weight - Statistical.max(results)
        elif type == InstanceType.MIN_WEIGHT_HARD:
            results = [result_default, result_max_profit, result_max_profit_per_weight_unit]
            return Statistical.min(results) - result_min_weight
        elif type == InstanceType.MAX_PROFIT_EASY:
            results = [result_default, result_min_weight, result_max_profit_per_weight_unit]
            return result_max_profit - Statistical.max(results)
        elif type == InstanceType.MAX_PROFIT_HARD:
            results = [result_default, result_min_weight, result_max_profit_per_weight_unit]
            return Statistical.min(results) - result_max_profit
        elif type == InstanceType.MAX_PROFIT_WEIGHT_EASY:
            results = [result_default, result_min_weight, result_max_profit]
            return result_max_profit_per_weight_unit - Statistical.max(results)
        elif type == InstanceType.MAX_PROFIT_WEIGHT_HARD:
            results = [result_default, result_min_weight, result_max_profit]
            return Statistical.min(results) - result_max_profit_per_weight_unit
        elif type == InstanceType.MAX_VARIANCE:
            results = [result_default, result_min_weight, result_max_profit, result_max_profit_per_weight_unit]
            return Statistical.stdev(results)
        elif type == InstanceType.MIN_VARIANCE:
            results = [result_default, result_min_weight, result_max_profit, result_max_profit_per_weight_unit]
            return -1 * Statistical.stdev(results)
        elif type == InstanceType.PAIRED_DEF_MAXP:
            max = Statistical.max([result_default, result_max_profit,
                                   result_max_profit_per_weight_unit, result_min_weight])
            result_default = result_default / max
            result_max_profit = result_max_profit / max
            result_max_profit_per_weight_unit = result_max_profit_per_weight_unit / max
            result_min_weight = result_min_weight / max
            return (_min(result_default, result_max_profit)
                    - _max(result_max_profit_per_weight_unit, result_min_weight)
                    - lambda_ * abs(result_default - result_max_profit))
        elif type == InstanceType.PAIRED_DEF_MAXPW:
            max = Statistical.max([result_default, result_max_profit,
                                   result_max_profit_per_weight_unit, result_min_weight])
            result_default = result_default / max
            result_max_profit = result_max_profit / max
            result_max_profit_per_weight_unit = result_max_profit_per_weight_unit / max
            result_min_weight = result_min_weight / max
            return (_min(result_default, result_max_profit_per_weight_unit)
                    - _max(result_max_profit, result_min_weight)
                    - lambda_ * abs(result_default - result_max_profit_per_weight_unit))
        elif type == InstanceType.PAIRED_DEF_MINW:
            max = Statistical.max([result_default, result_max_profit,
                                   result_max_profit_per_weight_unit, result_min_weight])
            result_default = result_default / max
            result_max_profit = result_max_profit / max
            result_max_profit_per_weight_unit = result_max_profit_per_weight_unit / max
            result_min_weight = result_min_weight / max
            return (_min(result_default, result_min_weight)
                    - _max(result_max_profit, result_max_profit_per_weight_unit)
                    - lambda_ * abs(result_default - result_min_weight))
        elif type == InstanceType.PAIRED_MAXP_MAXPW:
            max = Statistical.max([result_default, result_max_profit,
                                   result_max_profit_per_weight_unit, result_min_weight])
            result_default = result_default / max
            result_max_profit = result_max_profit / max
            result_max_profit_per_weight_unit = result_max_profit_per_weight_unit / max
            result_min_weight = result_min_weight / max
            return (_min(result_max_profit, result_max_profit_per_weight_unit)
                    - _max(result_default, result_min_weight)
                    - lambda_ * abs(result_max_profit - result_max_profit_per_weight_unit))
        elif type == InstanceType.PAIRED_MAXP_MINW:
            max = Statistical.max([result_default, result_max_profit,
                                   result_max_profit_per_weight_unit, result_min_weight])
            result_default = result_default / max
            result_max_profit = result_max_profit / max
            result_max_profit_per_weight_unit = result_max_profit_per_weight_unit / max
            result_min_weight = result_min_weight / max
            return (_min(result_max_profit, result_min_weight)
                    - _max(result_default, result_max_profit_per_weight_unit)
                    - lambda_ * abs(result_max_profit - result_min_weight))
        elif type == InstanceType.PAIRED_MAXPW_MINW:
            max = Statistical.max([result_default, result_max_profit,
                                   result_max_profit_per_weight_unit, result_min_weight])
            result_default = result_default / max
            result_max_profit = result_max_profit / max
            result_max_profit_per_weight_unit = result_max_profit_per_weight_unit / max
            result_min_weight = result_min_weight / max
            return (_min(result_max_profit_per_weight_unit, result_min_weight)
                    - _max(result_default, result_max_profit)
                    - lambda_ * abs(result_max_profit_per_weight_unit - result_min_weight))
        else:
            # MARKOVITZ_EASY and MARKOVITZ_HARD are declared but not implemented in
            # the Java version either, so they fall into this branch.
            print("The option is not recognized by the system.", file=sys.stderr)
            print("The system will halt.", file=sys.stderr)
            sys.exit(1)
        return 0


def _min(a: float, b: float) -> float:
    """Equivalent of ``Math.min(double, double)``."""
    return a if a < b else b


def _max(a: float, b: float) -> float:
    """Equivalent of ``Math.max(double, double)``."""
    return a if a > b else b
